framework/src/vectorpostprocessors/WorkBalance.C

```diff
@@ -0,0 +1,225 @@
+    for (decltype(n_sys) sys = 0; sys < n_sys; sys++)
+    {
+    */
+
+    // For MOOSE, system 0 is the nonnlinear system - that's what we care about here
+    // I've left the other code commented out here because I might change my mind in a little while
+    // and add back the ability to set the system or compute over all
+    unsigned int sys = 0;
+
+    auto n_vars = elem->n_vars(sys);
+
+    for (decltype(n_vars) var = 0; var < n_vars; var++)
+      _local_num_dofs += elem->n_dofs(sys, var);
+
+    //}
+  }
+
+  virtual void onInternalSide(const Elem * elem, unsigned int side) override
+  {
+    if (elem->neighbor(side)->processor_id() != _this_pid)
+    {
+      _local_num_partition_sides++;
+
+      // Build the side so we can compute its volume
+      auto side_elem = elem->build_side(side);
+      _local_partition_surface_area += side_elem->volume();
+    }
+  }
+
+  void join(const WBElementLoop & y)
+  {
+    _local_num_elems += y._local_num_elems;
+    _local_num_dofs += y._local_num_dofs;
+    _local_num_partition_sides += y._local_num_partition_sides;
+    _local_partition_surface_area += y._local_partition_surface_area;
+  }
+
+  dof_id_type _local_num_elems;
+  dof_id_type _local_num_dofs;
+  dof_id_type _local_num_partition_sides;
+  Real _local_partition_surface_area;
+
+  processor_id_type _this_pid;
+};
+
+class WBNodeLoop : public ThreadedNodeLoop<ConstNodeRange, ConstNodeRange::const_iterator>
+{
+public:
+  WBNodeLoop(FEProblemBase & fe_problem)
+    : ThreadedNodeLoop<ConstNodeRange, ConstNodeRange::const_iterator>(fe_problem),
+      _local_num_nodes(0),
+      _local_num_dofs(0)
+  {
+  }
+
+  WBNodeLoop(ThreadedNodeLoop & x, Threads::split split)
+    : ThreadedNodeLoop<ConstNodeRange, ConstNodeRange::const_iterator>(x, split),
+      _local_num_nodes(0),
+      _local_num_dofs(0)
+  {
+  }
+
+  virtual void onNode(ConstNodeRange::const_iterator & node_it)
+  {
+    auto & node = *(*node_it);
+
+    _local_num_nodes++;
+
+    unsigned int sys = 0;
+
+    auto n_vars = node.n_vars(sys);
+
+    for (decltype(n_vars) var = 0; var < n_vars; var++)
+      _local_num_dofs += node.n_dofs(sys, var);
+  }
+
+  void join(WBNodeLoop & y)
+  {
+    _local_num_nodes += y._local_num_nodes;
+    _local_num_dofs += y._local_num_dofs;
+  }
+
+  dof_id_type _local_num_nodes;
+  dof_id_type _local_num_dofs;
+};
+
+} // End of blank namespace
+
+void
+WorkBalance::execute()
+{
+  auto & mesh = _fe_problem.mesh();
+
+  // Get all of the Elem info first
+  auto wb_el = WBElementLoop(mesh);
+
+  Threads::parallel_reduce(*mesh.getActiveLocalElementRange(), wb_el);
+
+  _local_num_elems = wb_el._local_num_elems;
+  _local_num_dofs = wb_el._local_num_dofs;
+  _local_num_partition_sides = wb_el._local_num_partition_sides;
+  _local_partition_surface_area = wb_el._local_partition_surface_area;
+
+  // Now Node info
+  auto wb_nl = WBNodeLoop(_fe_problem);
+
+  Threads::parallel_reduce(*mesh.getLocalNodeRange(), wb_nl);
```

Comment:
Hmm, on one hand you aren't really after efficiency (you are doing two complete sweeps over the mesh when you could just do them together). On the other hand you went through the trouble of creating two custom threaded loops (when we don't run threads very often). Are you sure this is a better design than just gather both stats at once deriving from an existing looping type?

Reply:
Just asking the question. I just saw a bunch of new loops and wanted to make sure that this was the right choice here. Sounds like you've thought about it.